### PR TITLE
first version of the provider

### PR DESCRIPTION
**Pull request by dspasibenko**
the provider allows to control pool of chunks that can be used for reading data. The interface is not stabilitzed, so easily can be changed in the future...

### CODE REVIEW

File: pkg/storage/chunkfs/provider.go [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+	select {
+	case <-ctx.Done():
+		p.ReleaseChunk(c)
+		return nil, ctx.Err()
+	case <-p.closedCh:
+		p.ReleaseChunk(c)
+		return nil, errors.ErrClosed
+	case <-c.neverOpened:
+	}
+	return c, nil
```
Comment (kbabushkin) [inline pkg/storage/chunkfs/provider.go:82]: q: can we get a closed chunk? I mean, the opener hasn't been yet able to `c.Open()` it but we already tried to `c.Append`...

**Review by kbabushkin — APPROVED** [review]
🪨

File: pkg/storage/chunkfs/provider.go [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+		// borrow one counter
+		p.active++
+		last := len(p.toOpenList) - 1
+		c := p.toOpenList[last]
+		p.toOpenList[last] = nil
+		p.toOpenList = p.toOpenList[:last]
+
+		if _, ok := p.standBy.Get(c.id); ok {
+			p.logger.Debugf("found the chunk=%v in the toOpenList, but all requestors are gone, will not open it now.", c)
+			// the chunk was requested to be opened, but all clients, who wanted to work with it, gone. Remove is from the
```
[the remark this answers is not in the document]
Comment (dspasibenko) [reply pkg/storage/chunkfs/provider.go:235]: ok

File: pkg/storage/chunkfs/provider.go [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+	lock   sync.Mutex
+	cond   *sync.Cond
+	logger logging.Logger
+	dir    string
+	chunks map[string]*Chunk
+	active int
+	maxFDs int
+	// standBy is the collection of opened Chunks that maybe removed
+	standBy *iterable.Map[string, *Chunk]
+	// otOpenList the list of chunks that have to be opened
```
[the remark this answers is not in the document]
Comment (dspasibenko) [reply pkg/storage/chunkfs/provider.go:42]: yes

File: pkg/storage/chunkfs/provider.go [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+		if len(p.toOpenList) == 0 {
+			// waiting till a new chunk be requested to be opened
+			p.cond.Wait()
+			continue
+		}
+
+		if p.active == p.maxFDs {
+			// the number of open chunks hits the limit
+			if p.standBy.Len() > 0 {
+				// some are opened and not used, let's close the lest used one
```
[the remark this answers is not in the document]
Comment (dspasibenko) [reply pkg/storage/chunkfs/provider.go:210]: cool

File: pkg/storage/chunkfs/provider.go (hunk 1 above)
Comment (dspasibenko) [reply pkg/storage/chunkfs/provider.go:82]: it was not an idea, the closed chunk cannot be returned by the function. `c.neverOpened` is not closed before the chunk is opened.

**Pull request merged**